Library/Homebrew/rubocops/text.rb

```diff
@@ -154,6 +148,21 @@ def audit_formula(_node, _class_node, _parent_class_node, body_node)
             problem "`env :std` in homebrew/core formulae is deprecated"
           end
         end
+
+        # Check whether value starts with the formula name and then a "/", " " or EOS
+        def starts_with_formula_name?(value)
+          value.match?(%r{#{Regexp.escape(@formula_name)}(/| |$)})
+        end
+
+        # Find "#{share}/foo"
+        def_node_search :share_formula_name_dstr, <<~EOS
+          $(dstr (begin (send nil? :share)) (str #starts_with_formula_name?))
```

Comment:
Where has the `/` gone?

Reply:
Hmm, good catch. I think this should have made the tests fail? I'll look into it

Reply:
Ah, looks like I forgot to add the `^` to the start of the regex so I just got lucky with the tests. I'll add a new test that would fail in this situation.

---

What I wanted to do for these methods is to have the definition be:

```ruby
def_node_search :share_formula_name_dstr, <<~EOS
    $(dstr (begin (send nil? :share)) (str %1))
EOS
```

and call it using:

```ruby
share_formula_name_dstr(body_node, %r{^/#{Regexp.escape(@formula_name)}(/| |$})
```

passing the regex with the formula name in for `%1`. This would avoid needing the extra method to check whether the string starts with the formula name.

This worked fine when passing a string directly, but I couldn't get it to work with a regex (even though the [docs](https://docs.rubocop.org/rubocop-ast/node_pattern.html#for-arguments) indicate that a regex should work). Any ideas?

Reply:
No idea but I think I’d prefer the `starts_with_formula_name?` method.
It’s perfectly fine and maintainable.

Reply:
I updated the `starts_with_formula_name?` method to have an optional `prefix`:

https://github.com/Homebrew/brew/blob/d8fb850fa9fd535a0e08149b4faed2a1ce73a44a/Library/Homebrew/rubocops/text.rb#L152-L155

So now the `def_node_search` can pass a slash to prefix for the `dstr` check:

https://github.com/Homebrew/brew/blob/d8fb850fa9fd535a0e08149b4faed2a1ce73a44a/Library/Homebrew/rubocops/text.rb#L157-L160

This makes the node pattern a little more difficult to read, but I thought it was a better solution than adding an optional `/` at the start of the regex.

Another solution would be to have two methods (maybe `starts_with_formula_name?` and `starts_with_slash_formula_name?`) but I felt that was unnecessary. Any preference?

Reply:
I’d say we should stick to a single method.
You can make it easier to understand by passing `@formula` and renaming the method accordingly.

For example:

```
str #path_starting_with?(@formula)
str #path_starting_with?("/#{@formula}")
```

Reply:
Oh yeah, I like that. Good idea.

Reply:
`@formula_name` isn't set when `def_node_search` runs, but doing this, instead, works:

```ruby
def_node_search :share_formula_name_dstr, <<~EOS
  $(dstr (begin (send nil? :share)) (str #path_starts_with?(%1)))
EOS
```

```ruby
share_formula_name_dstr(body_node, "/#{@formula_name}") do ...
```

Reply:
Thoughts on method names?

I think `share_path_starts_with` and `dstr_share_path_starts_with` might be clearer.

`dstr` refers to the node type for interpolated strings, but maybe `interpolated_share_path_starts_with` is a better name?

Reply:
I think `share_path_starts_with` and `interpolated_share_path_starts_with` are perfectly readable and work just fine.